### ISSUE TITLE
TypeError: Cannot read properties of undefined in Content-Type Builder and Uploads pages

### ISSUE THREAD

**Issue by Sorok-Dva** [strapi/strapi#12837]
## Bug report
Hello,

I've an issue in the "Content-Type Build" page with the error `Cannot read properties of undefined (reading 'startSection')`
and the "uploads" page (under the content-type build) with the error `Cannot read properties of undefined (reading 'allPermissions')`

### Describe the bug
 When I'm going on the url `127.0.0.1/admin/plugins/content-type-builder` i get a blank page (here black because of my browser dark mode) with this error in the console
 
(c.f Screen 1)

Also the same problem but with another property in the uploads pages (`127.0.0.1/admin/plugins/upload`)

(c.f Screen 2)

I tried with version 4.0.5, 4.1.0, 4.1.1, 4.1.2 & 4.1.3
Each time I removed the package-lock.json, removed the `build` and `.cache` folders, and then start the server and I always have the same problem on the both pages.

### Steps to reproduce the behavior

1. Go to admin panel
2. Click on Content-Type Builder
4. See blank page & error
5. Go to admin panel index by refreshing or opening a new tab
6. Click on "Media library"
7. See blank page & error

### Expected behavior

Accessing both content type builder and media library page without errors in console and blank page.

### Screenshots

Screen 1 : ![image](https://user-images.githubusercontent.com/15158794/158209435-8381f34f-3cb6-4b1d-90b5-97bceda667e3.png)
Screen 2: ![image](https://user-images.githubusercontent.com/15158794/158210577-507c18d0-4137-43ae-b75a-55af512dc42d.png)

### System

- Node.js version: v14.19.0
- NPM version: v14.19.0
- Strapi version: Tested on 4.0.5, 4.1.0, 4.1.1, 4.1.2 & 4.1.3
- Database: Mysql 8.0.28
- Operating system: Ubuntu 18.04.6 LTS (GNU/Linux 4.15.0-171-generic x86_64)

### Additional context

The problem appears after I got for an unknown reason a deletion of the content of the strapi folder, after reverting it with git, the problem was here.

**Comment by imiskolee** [comment]
same issues now

**Comment by maxpropp** [comment]
I am having a similar issue, but in my case scenario I can enter the content type builder. But this error appears when I upload something in the media field from a content type. The page becomes blank. If I refresh, everything will be fine. The file is successfuly uploaded, but everytime something is uploaded it just goes blank.

**Comment by talazenkot** [comment]
This is a critical bug... experiencing it too

**Comment by gu-stav** [comment]
Hello everyone. Thank you for reporting this issue. I tried to reproduce it with a clean installation of a recent version of Strapi, but could not.

Could you provide me with some clear reproduction steps with a clean installation? In case you have any plugins installed, I would also appreciate it a lot if you could share the content of your package.json file.

**Comment by maxpropp** [comment]
@gu-stav sorry I can't do that right now. But I do have an idea of what could be happening (at least in my case, not sure if that is the same as @Sorok-Dva and @talazenkot ). 

It seems that I am limiting the Media type into files, and I was uploading an image, so it was doing this weird behavior. It does not actually block me from uploading, but it goes blank after uploading (it still uploads successfuly).

**Comment by tuxuuman** [comment]
I have the same problem on the **content-type-builder** page.  Appeared after upgrading from 4.1.2 to 4.1.5 and removing **.cache** and **build** folders
![image](https://user-images.githubusercontent.com/17300270/158821404-3b789198-4de6-4b59-8cfc-be2e360c07a2.png)

**Comment by gu-stav** [comment]
@maxpropp Thanks, that was helpful. I was able to reproduce your issue. Do you mind opening another issue about this case? I believe it's a different one described in this issue here.

**Comment by talazenkot** [comment]
> I have the same problem. Appeared after upgrading from 4.1.2 to 4.1.5 and removing **.cache** and **build** folders. Rolling back to previous version did not help

@tuxuuman I had the exact same issue. Rolling back to 4.1.3 and deleting both `.cache` and `build` folders **again** solved the issue for me, and now it doesn't break. Anyway, it's still a critical issue because unless I'm missing something, it prevents from upgrading to 4.1.5 for now.

**Comment by gu-stav** [comment]
@talazenkot I'd love to start fixing this issue since it seems to affect quite a few users - currently, I'm not able to reproduce it on a clean installation. It would be very helpful if you could provide a reproduction if you know how. 🙏🏼

**Comment by tiennm-galaxt** [comment]
The same issue.
I had this issue when using an older strapi version than the latest version.

My package.json
```
{
  "name": "name",
  "private": true,
  "version": "0.1.4",
  "description": "A Strapi application",
  "scripts": {
    "develop": "strapi develop",
    "start": "strapi start",
    "build": "strapi build",
    "strapi": "strapi",
    "ui": "npm run build --develop && npm run develop"
  },
  "dependencies": {
    "@strapi/plugin-documentation": "4.1.3",
    "@strapi/plugin-i18n": "4.1.3",
    "@strapi/plugin-users-permissions": "4.1.3",
    "@strapi/strapi": "4.1.3",
    "directory-tree": "3.2.2",
    "pg": "8.6.0"
  },
  "author": {
    "name": "A Strapi developer"
  },
  "strapi": {
    "uuid": "my-uuid"
  },
  "engines": {
    "node": ">=12.x.x <=16.x.x",
    "npm": ">=6.0.0"
  },
  "license": "MIT"
}
```

I get an error when I build docker file production. But when I upgrade to 4.1.5 there is no more error on build
That repeats when strapi releases a new tag on github. 

Example: 
- I'm using 4.1.3, latest version is 4.1.5 => ERROR.
- I'm using 4.1.5, latest version is 4.1.5 => NOT ERROR.

```
FROM node:16-alpine
ENV NODE_ENV=production

WORKDIR /app
COPY ./ ./

RUN npm i
RUN npm run build --production
CMD ["npm", "start"]
```

**Comment by mariansimecek** [comment]
Any updates or way to fix? I can´t use updated version of strapi.

**Comment by linbudu599** [comment]
Waiting for solutions...

**Comment by tiennm-galaxt** [comment]
> This seems to be a problem reletaed to package manager? When using pnpm, no version can work correctly, as using yarn works correctly for every version?
[SOLVE]
Thanks for solution. I no longer get errors when I switch to yarn
My Dockerfile.
```
FROM node:16-alpine
ENV NODE_ENV=production

WORKDIR /app
COPY ./ ./

RUN yarn
RUN yarn run build
CMD ["yarn", "start"]
```

**Comment by linbudu599** [comment]
> > This seems to be a problem reletaed to package manager? When using pnpm, no version can work correctly, as using yarn works correctly for every version?
> 
> [SOLVE] Thanks for solution. I no longer get errors when I switch to yarn My Dockerfile.
> 
> ```
> FROM node:16-alpine
> ENV NODE_ENV=production
> 
> WORKDIR /app
> COPY ./ ./
> 
> RUN yarn
> RUN yarn run build
> CMD ["yarn", "start"]
> ```

I solved this too by switching to yarn, but I'm developing in pnpm workspace, so this is still blocking me :-(

**Comment by soupette** [comment]
@linbudu599 can you detail your setup correctly and all the commands in order to reproduce the issue. Since it seems to be related to the package manager I am thinking that the issue might be related to having the @strapi/helper-plugin package dependency in the following plugins:
- @strapi/plugin-content-type-builder
- @strapi/plugin-email
- @strapi/plugin-upload
- @strapi/plugin-documentation
- @strapi/plugin-users-permissions


Can you try to remove the dependency from theses packages using [patch-package](https://www.npmjs.com/package/patch-package) and let me know if it works correctly after reinstalling everything?

I'll then make a PR in the project to remove then completely.

**Comment by neilpoulin** [comment]
i ran into these exact error messages when upgrading from strapi `4.1.5` to `4.1.7`. I was finally able to resolve it by removing all of the following and doing a fresh install (I did not reset my database) `node_modules`, `.cache`, `build`, `package-lock.json`. 

I then did `npm install` and then `npm build` and finally `npm develop`. I can now access my media library and content builder pages like as normal. 

Also a note in case it helps anyone else: i may have screwed myself up by initially installing the strapi update using a different version of node (16.x, which is not supported by strapi), so it may have put my package.json into a bad state.

**Comment by TheBit** [comment]
@linbudu599 there is a fork for `patch-package` which seems to have some sort of pnpm support (without workspaces): https://github.com/ds300/patch-package/pull/302
Have you heard of it? Have you maybe tried it?

**Comment by rowphant** [comment]
I had the same loading issues with alle settings pages "USERS & PERMISSIONS PLUGIN" after I upgraded Strapi from 4.1.3 to 4.1.7
My working solution was quite easy after I took a look in the package.json:
Upgrade the package "@strapi/plugin-users-permissions" to the latest version (4.1.7) as well.

Hope this is helpful for someone.

**Comment by aquadesk** [comment]
```
rm build -rf 
yarn dev
```

This solved it for me on the version 4.1.8
Thanks

**Comment by derrickmehaffy** [comment]
Marking as closed as the issue was a cached admin build

**Comment by thonatos** [comment]
the way to reproduce

1. create a pnpm workspace
2. init strapi with name 'demo' at packages/apps/demo
3. install deps with `pnpm i --filter demo...`
4. run build under demo app with `pnpm run build`
5. copy deps to demo's package.json from package.json under .cache
6. open the browser after building admin

**Comment by TheBit** [comment]
@thonatos the same problem can be reproduced in a Rush.js monorepo with pnpm as the selected package manager.
Because of this, I can't move Strapi to Rush monorepo to store its sources together with the front-ends which consume Strapi (to get benefits of performing breaking changes in a single MR (Strapi + FE)... :(

**Comment by mortenmoulder** [comment]
What I did to reproduce a white screen with the `Cannot read properties of undefined (reading 'startSection')` message (as explained in my issue #13469), as well as a way to make it work again:

1. Create new project with `npx create-strapi-app my-app-name --template blog --quickstart`
2. Navigate to http://localhost:1337/admin and create an admin account
3. Go to the Content-Type Builder and find the Article collection type: http://localhost:1337/admin/plugins/content-type-builder/content-types/api::article.article (https://i.imgur.com/9LNSaqh.png)
4. Remove Media, Quote, and Slider from the blocks section and press save: https://i.imgur.com/FNBiFaN.mp4
5. Stop the application by pressing CTRL+C in the terminal
6. Run `npm run build`
7. Run `npm run develop`
8. Go to the Content Manager and open the Article collection type: https://i.imgur.com/3WHyeub.png
9. Click on an article (for example "The internet's Own boy" with ID = 1)
10. Go to the Content-Type Builder, find the Article collection type, and add the Media, Quote, and Slider to the blocks section again
11. Go to the article from step 8-9 and they work again
12. Remove the Media, Quote, and Slider from the article you opened
13. Repeat step 3-4
14. Go to the article again and now it works again, but the other articles (that still have data in those components we just removed) are broken

I'm pretty sure you can skip step 6 and 7, because the problem seem clear to me: There's still data in the component you just removed. The API confirms there's still data in the content-type, even though I removed the components. 

@derrickmehaffy Can you confirm?

**Comment by derrickmehaffy** [comment]
Thank you for these reproduction steps :pray: :heart: let me take a look

**Comment by derrickmehaffy** [comment]
I can confirm this given the steps, will reopen

**Comment by kibblerz** [comment]
https://github.com/strapi/strapi/pull/13513

I created this PR, which fixes the issue for me. (While also adding Yarn PnP support without loose mode). I was using yarn though. Currently the PR won't work with npm due to @strapi/icons requesting an alpha version of the design system. With yarn PnP, this issue was resolved for me

**Comment by gu-stav** [comment]
Hello everybody, we have decided to prioritize this issue. 🚀 

It is not yet fully clear to me what actually happens. What I can say: you are right: it doesn't have anything to do with your plugin or a missing method, but with the way we bundle the whole application.

Could you do me a favor and try to replace the following file:

`node_modules/@strapi/admin/webpack.alias.js`

with

```js
'use strict';

const path = require('path');

const alias = [
  'whatwg-fetch',
  '@fortawesome/fontawesome-svg-core',
  '@fortawesome/free-solid-svg-icons',
  '@strapi/design-system',
  '@strapi/icons',
  'date-fns',
  'formik',
  'history',
  'immer',
  'lodash',
  'moment',
  'qs',
  'react',
  'react-copy-to-clipboard',
  'react-dnd',
  'react-dnd-html5-backend',
  'react-dom',
  'react-error-boundary',
  'react-fast-compare',
  'react-helmet',
  'react-is',
  'react-intl',
  'react-redux',
  'react-router',
  'react-router-dom',
  'react-virtualized',
  'react-select',
  'redux',
  'reselect',
  'styled-components',
  'yup',
];

module.exports = alias.reduce(
  (acc, curr) => {
    acc[`${curr}$`] = require.resolve(curr);

    return acc;
  },
  {
    'react-select/animated': require.resolve('react-select/animated'),
    'react-select/async': require.resolve('react-select/async'),
    'react-select/async-creatable': require.resolve('react-select/async-creatable'),
    'react-select/base': require.resolve('react-select/base'),
    'react-select/creatable': require.resolve('react-select/creatable'),
    '@strapi/helper-plugin': require.resolve('@strapi/helper-plugin'),
    ee_else_ce: path.resolve(__dirname),
  }
);
```

and try to re-run `yarn build` to see if that fixes the issue for you?

Refs: https://github.com/strapi/strapi/issues/13222

**Comment by gu-stav** [comment]
@arpadrozsa I agree and I believe the same. This is the main reason for the `webpack.alias.js`: it tries to resolve all the different dependencies from the global `node_modules` directory instead of the local ones. This is necessary because the `@strapi/helper-plugin` contains all global providers and if it would be resolved from different locations, we get different global states per include.

At the moment I'm testing on `yarn@1` and `npm@7` only, but I'll check if something changes using v8. Thanks!

**Comment by NonOrganicCreature** [comment]
![image](https://user-images.githubusercontent.com/73524976/179400268-00cf4e4f-106a-4e2b-86bd-d4c67937ab2b.png)
I also change before build webpack.alias.js
Dockerfile
```FROM node:16.14-alpine
WORKDIR /app
RUN corepack enable
COPY package.json yarn.lock ./
RUN yarn install
COPY . .
RUN cp -f webpack.alias.js /app/node_modules/@strapi/admin/webpack.alias.js
RUN yarn build
CMD ["yarn", "start"]```

- Node.js version: 16.14-alpine
- YARN version: 1.2.24
- Strapi version: 4.2.2
- Database: postgresql
- Operating system: Ubuntu 18.04

Steps to reproduce are similar to the first post

**Comment by bartoszkrawczyk2** [comment]
This issue is still not resolved. I'm on version 4.3.4 and I get `TypeError: Cannot read properties of undefined (reading 'icon')` When opening any article in Content Manager. Removed cache and node_modules, reinstalled everything but the issue still persist.

**Comment by soupette** [comment]
@gu-stav if it still related to the resolution of the helper plugin you might need to create singleton for each context provider to instantiate them only once.

**Comment by gu-stav** [comment]
@soupette Do you mean instead of doing:

```js
import { createContext } from 'react';

const TrackingContext = createContext();

export default TrackingContext;
```

we'd need to export?

```js
import { createContext } from 'react';

export const initContext = () => {
  const TrackingContext = createContext();

  return TrackingContext;
}
```

Or do you have in mind we need to "cache" the context?

```js
import { createContext } from 'react';

let TrackingContext;

export const initContext = () => {
  if (!TrackingContext) {
    TrackingContext = createContext();
  }

  return TrackingContext;
}
```

**Comment by soupette** [comment]
eslint is going to complain if you using hook inside a statement, you can check out this [article](https://dev.to/kardysm/singleton-like-context-for-shared-components-managent-4112) how to create a singleton.

I have also seen this pattern:

```js
const createSingleton = Compo => {
  const SingletonContext = createContext({ hasBeenInstantiated: false })
  
    return function Singleton({ children, value }) {
    const singleton = useContext(SingletonContext)

    if (!singleton.hasBeenInstantiated || (singleton.value !== value && value !== undefined)) {
      return (
        <SingletonContext.Provider value={{ hasBeenInstantiated: true, value }}>
          <Compo value={value}>{children}</Compo>
        </SingletonContext.Provider>
      )
    }

    return children
  }
}

const SingletonProvider = createSingleton(MyProvider)
```

**Comment by gu-stav** [comment]
Hello everyone 👋🏼 

We believe we found a better fix for the issue and released an experimental version to be able to test that. Could you do us a big favor and check if you still run into this problem when using:

```
npx create-strapi-app@0.0.0-b4eaa97842f53ed5611c878ecef9654b7bde1c11 my-app --quickstart
```

If you have a running Strapi application, you can update its dependencies to:

```
dependencies: {
  "@strapi/strapi": "0.0.0-b4eaa97842f53ed5611c878ecef9654b7bde1c11"
}

**Comment by TheBit** [comment]
@gu-stav 
Sorry for the slow response, overwhelmed with my current duties on a job :(
My setup requires a lot of hassle, so it can't be checked quickly (I'm trying to move my Strapi from separate repo into Rush.js monorepo with pnpm).
But I will definitely check as soon I find the time (hopefully soon).
Thanx for pushing this!
